src/openrct2/ride/TrackStyle.cpp

```diff
@@ -0,0 +1,105 @@
+    GetTrackPaintFunctionMineRide,                 // mineRide
+    GetTrackPaintFunctionMineTrainRC,              // mineTrainCoaster
+    GetTrackPaintFunctionMiniGolf,                 // miniGolf
+    GetTrackPaintFunctionMiniHelicopters,          // miniHelicopters
+    GetTrackPaintFunctionMiniRC,                   // miniRollerCoaster
+    GetTrackPaintFunctionMiniSuspendedRC,          // miniSuspendedCoaster
+    GetTrackPaintFunctionMiniatureRailway,         // miniatureRailway
+    GetTrackPaintFunctionMonorail,                 // monorail
+    GetTrackPaintFunctionMonorailCycles,           // monorailCycles
+    GetTrackPaintFunctionMotionsimulator,          // motionSimulator
+    GetTrackPaintFunctionMultiDimensionRC,         // multiDimensionRollerCoaster
+    GetTrackPaintFunctionMultiDimensionRC,         // multiDimensionRollerCoasterInverted
+    GetTrackPaintFunctionObservationTower,         // observationTower
+    GetTrackPaintFunctionReverseFreefallRC,        // reverseFreefallCoaster
+    GetTrackPaintFunctionReverserRC,               // reverserRollerCoaster
+    GetTrackPaintFunctionRiverRapids,              // riverRapids
+    GetTrackPaintFunctionRotoDrop,                 // rotoDrop
+    GetTrackPaintFunctionShop,                     // shop
+    GetTrackPaintFunctionSideFrictionRC,           // sideFrictionRollerCoaster
+    OpenRCT2::SingleRailRC::GetTrackPaintFunction, // singleRailRollerCoaster
+    GetTrackPaintFunctionSpaceRings,               // spaceRings
+    GetTrackPaintFunctionSpiralSlide,              // spiralSlide
+    GetTrackPaintFunctionSplashBoats,              // splashBoats
+    GetTrackPaintFunctionStandUpRC,                // standUpRollerCoaster
+    GetTrackPaintFunctionWildMouse,                // steelWildMouse
+    GetTrackPaintFunctionSteeplechase,             // steeplechase
+    GetTrackPaintFunctionSubmarineRide,            // submarineRide
+    GetTrackPaintFunctionSuspendedMonorail,        // suspendedMonorail
+    GetTrackPaintFunctionSuspendedSwingingRC,      // suspendedSwingingCoaster
+    GetTrackPaintFunctionSwingingInverterShip,     // swingingInverterShip
+    GetTrackPaintFunctionSwingingShip,             // swingingShip
+    GetTrackPaintFunctionTopspin,                  // topSpin
+    GetTrackPaintFunctionTwist,                    // twist
+    GetTrackPaintFunctionTwisterRC,                // twisterRollerCoaster
+    GetTrackPaintFunctionVirginiaReel,             // virginiaReel
+    GetTrackPaintFunctionWaterRC,                  // waterCoaster
+    GetTrackPaintFunctionWoodenRC,                 // woodenRollerCoaster
+    GetTrackPaintFunctionWoodenWildMouse,          // woodenWildMouse
+};
+
+TRACK_PAINT_FUNCTION GetTrackPaintFunction(TrackStyle trackStyle, OpenRCT2::TrackElemType trackType)
+{
+    auto cast = static_cast<uint8_t>(trackStyle);
+    if (cast >= std::size(kPaintFunctionMap))
+        return nullptr;
```

Comment:
Is this even a valid path and if so why? I feel like there should be no way to ever have this function receive an invalid TrackStyle value. We should always have such checks during import and not when its used, we should also make it guarantee its never null by returning a reference.

Reply:
The track style will be saved in the tile element very soon, at which point all bets are off and it cases like that _will_ happen.

As for returning nullptr vs. reference: this function _cannot_ return a reference. If you track down `kPaintFunctionMap[cast](trackType)`, you’ll notice that it will return a nullptr in the _majority_ of cases - there are 340 different TrackElemTypes and every coaster only implements a subset, returning `nullptr` from its corresponding `GetTrackPaintFunctionFooBar()` function (referenced in `kPaintFunctionMap[cast]`) in the other cases. The only way to change that is if all the track drawer would return a dummy function if it cannot draw the element, but I’m not sure if that is worth it.

Reply:
I don't think we should ever allow invalid values to be set by user or imported from file. I think having something like a dummy one is not that far off, we could treat 0/invalid as invisible track which is our dummy/NOP, on import we can simply check if its one of the accepted values and if not turn it invisible/dummy/NOP. The reason as for why a reference is that we eliminate branching which is not zero cost and for developers they don't have to check if the function can possibly return null. What do you think?